File: 2016/python/day2/day2.py
from common import io, util


def run():
    print("Part 1 result:", part1())


def part1():
    return [list(reversed(list(get_key(x))))[0] for x in parse_input("day2/input.txt")]


def parse_input(path):
    return [list(x) for x in io.read_file(path).splitlines(False)]


def get_key(path):
    ptr, keys = [1, 1], [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
    for ch in path:
        ptr = [util.clamp(x + y, 0, 2) for (x, y) in zip(ptr, DIRECTIONS[ch])]
        yield keys[ptr[1]][ptr[0]]


DIRECTIONS = {'L': [-1, 0], 'U': [0, -1], 'R': [1, 0], 'D': [0, 1]}

File: 2015/go/day2/day2.go
package day2

import (
	"fmt"
	"sort"
	"strconv"
	"strings"

	"../common"
)

// Day2 : Solutions
func Day2() {
	fmt.Println("Part 1 result:", part1())
}

func part1() int {
	pres := parseInput(common.ReadFile("day2/input.txt"))

	paper := 0
	for _, p := range pres {
		paper += p.paperRequired()
	}

	return paper
}

type present struct {
	x, y, z int
}

func (p present) paperRequired() int {
	return (2 * p.x * p.y) +
		(2 * p.y * p.z) +
		(2 * p.z * p.x) +
		p.smallestSideArea()
}

func (p present) smallestSideArea() int {
	dims := []int{p.x, p.y, p.z}
	sort.Ints(dims)

	return dims[0] * dims[1]
}

func parsePresent(s string) present {
	comp := strings.Split(s, "x")
	getVal := func(s string) int {
		val, err := strconv.ParseInt(s, 10, 0)
		if err != nil {
			panic("Conversion failed")
		}
		return int(val)
	}
	return present{getVal(comp[0]), getVal(comp[1]), getVal(comp[2])}
}

func parseInput(input string) []present {
	lines := common.GetLines(input)
	pres := []present{}

	for _, line := range lines {
		pres = append(pres, parsePresent(line))
	}

	return pres
}
